tiledb/api/src/datatype/arrow.rs

```diff
@@ -46,169 +53,667 @@ impl ArrowPrimitiveTypeNative for f64 {
+    }
+
+    pub fn is_exact(&self) -> bool {
+        matches!(self, Self::Exact(_))
+    }
+
+    pub fn into_inner(self) -> arrow::datatypes::DataType {
+        match self {
+            Self::Exact(arrow) => arrow,
+            Self::Inexact(arrow) => arrow,
+        }
     }
 }
 
-/// For an Arrow type, returns a TileDB type if the bits of the canonical input type match.
-/// If this returns Some(tdb_t), then values for tdb_t can be used in functions which expect
-/// arrow_dt and vice verse.
-pub fn tiledb_type_physical(
-    arrow_dt: &arrow_schema::DataType,
-) -> Option<Datatype> {
-    match *arrow_dt {
-        arrow_schema::DataType::Int8 => Some(Datatype::Int8),
-        arrow_schema::DataType::Int16 => Some(Datatype::Int16),
-        arrow_schema::DataType::Int32 => Some(Datatype::Int32),
-        arrow_schema::DataType::Int64 => Some(Datatype::Int64),
-        arrow_schema::DataType::UInt8 => Some(Datatype::UInt8),
-        arrow_schema::DataType::UInt16 => Some(Datatype::UInt16),
-        arrow_schema::DataType::UInt32 => Some(Datatype::UInt32),
-        arrow_schema::DataType::UInt64 => Some(Datatype::UInt64),
-        arrow_schema::DataType::Float32 => Some(Datatype::Float32),
-        arrow_schema::DataType::Float64 => Some(Datatype::Float64),
-        arrow_schema::DataType::Timestamp(
-            arrow_schema::TimeUnit::Second,
-            None,
-        ) => Some(Datatype::DateTimeSecond),
-        arrow_schema::DataType::Timestamp(
-            arrow_schema::TimeUnit::Millisecond,
-            None,
-        ) => Some(Datatype::DateTimeMillisecond),
-        arrow_schema::DataType::Timestamp(
-            arrow_schema::TimeUnit::Microsecond,
-            None,
-        ) => Some(Datatype::DateTimeMicrosecond),
-        arrow_schema::DataType::Timestamp(
-            arrow_schema::TimeUnit::Nanosecond,
-            None,
-        ) => Some(Datatype::DateTimeNanosecond),
-        arrow_schema::DataType::Time64(arrow_schema::TimeUnit::Microsecond) => {
-            Some(Datatype::TimeMicrosecond)
-        }
-        arrow_schema::DataType::Time64(arrow_schema::TimeUnit::Nanosecond) => {
-            Some(Datatype::TimeNanosecond)
-        }
-        _ => None, // TODO
+/*
+ * (Datatype::StringAscii, CellValNum::Var) does not have an exact analog in Arrow.
+ * Utf8 sounds pretty good, but we can't use it because Arrow validates Utf8 and
+ * tiledb does not. So we use `LargeList(UInt8)` instead.
+ * However, in tiledb StringAscii has several special accommodations which
+ * are not granted to UInt8. We must be able to invert back to StringAscii.
+ * We can do that by storing the exact input datatype on the arrow list field metadata.
+ */
+/// `arrow::datatypes::Field` metadata key for an exact `tiledb::Datatype` variant
+/// if there is no exact mapping from `tiledb::Datatype` to `arrow::datatypes::DataType`.
+pub const ARROW_FIELD_METADATA_KEY_TILEDB_EXACT_MATCH: &str =
+    "tiledb_exact_datatype";
```

Comment:
This logic is confusing me quite a bit. Are we tagging values that are an exact match and if so, why? I would expect an optional key/value that say "use this type instead" only when there's *not* an exact match. This sort of approach means we'd round trip everything we produced correctly while also transparently accepting anything we didn't produce (assuming we support it).